cis194/week2/tien/homework2/src/LogAnalysis.hs

```diff
@@ -1,26 +1,56 @@
 {-# OPTIONS_GHC -Wall #-}
+{-# LANGUAGE ViewPatterns #-}
 
 module LogAnalysis where
 
 import           Log
 import           Text.Read (readMaybe)
 
+-- Exercise 1 --
+-- parseMessage :: String -> LogMessage
+-- parseMessage logStr =
+--   case words logStr of
+--     ("I":ts:rest) ->
+--       case readMaybe ts :: Maybe Int of
+--         Just i -> LogMessage Info i (unwords rest)
+--         _      -> Unknown logStr
+--     ("W":ts:rest) ->
+--       case readMaybe ts :: Maybe Int of
+--         Just i -> LogMessage Warning i (unwords rest)
+--         _      -> Unknown logStr
+--     ("E":lv:ts:rest) ->
+--       case (readMaybe lv :: Maybe Int, readMaybe ts :: Maybe Int) of
+--         (Just i, Just j) -> LogMessage (Error i) j (unwords rest)
+--         _                -> Unknown logStr
+--     _ -> Unknown logStr
+-- Use ViewPatterns --
 parseMessage :: String -> LogMessage
-parseMessage logStr =
-  case words logStr of
-    ("I":ts:rest) ->
-      case readMaybe ts :: Maybe Int of
-        Just i -> LogMessage Info i (unwords rest)
-        _      -> Unknown logStr
-    ("W":ts:rest) ->
-      case readMaybe ts :: Maybe Int of
-        Just i -> LogMessage Warning i (unwords rest)
-        _      -> Unknown logStr
-    ("E":lv:ts:rest) ->
-      case (readMaybe lv :: Maybe Int, readMaybe ts :: Maybe Int) of
-        (Just i, Just j) -> LogMessage (Error i) j (unwords rest)
-        _                -> Unknown logStr
-    _ -> Unknown logStr
+parseMessage (words -> "I":(readMaybe -> Just ts):msg) =
+  LogMessage Info ts (unwords msg)
+parseMessage (words -> "W":(readMaybe -> Just ts):msg) =
+  LogMessage Warning ts (unwords msg)
+parseMessage (words -> "E":(readMaybe -> Just lvl):(readMaybe -> Just ts):msg) =
+  LogMessage (Error lvl) ts (unwords msg)
+parseMessage logStr = Unknown logStr
 
 parse :: String -> [LogMessage]
 parse = fmap parseMessage . lines
+
+-- Exercise 2 --
+constrcutTreeByTs ts1 ts2 log1 log2 left right
+  | ts1 > ts2 = Node left log2 (insert log1 right)
+  | ts1 < ts2 = Node (insert log1 left) log2 right
+
+insert :: LogMessage -> MessageTree -> MessageTree
+insert log (id -> Leaf) = Node Leaf log Leaf
+insert log1@(LogMessage (Error _) ts1 _) (Node left log2@(LogMessage (Error _) ts2 _) right) =
```

Comment:
The 3 pattern matches from this line onwards express the following scenarios:
- insert `Error` to a `Node` tagged by `Error`
- insert `Error` to a `Node` tagged by any `LogMessage`
- insert any `LogMessage` into a `Node` tagged by  any `LogMessage`
All 3 above has the same logic after match: `constrcutTreeByTs ts1 ts2 log1 log2 left right`
Does this means, we only need the last one? as in, when the log being inserted and the Node that the log is inserting to both have `LogMessage` as constructor in them.

Reply:
I think I tried to get ts1 and ts2 from different log messages as there are 2 different structures of the log messages.

Reply:
because we pattern match to get ONLY the timestamp from a message constructed by LogMessage, we can always ignore the message type.